commit 5ceee3de04e8bcb7048b8bca2c9728f89bf974de
Author: Michael Still <mikal@stillhq.com>

    [libvirt] Optionally enable SPICE debug logging.

    This makes it much easier to debug connectivity issues when they
    occur.

diff --git a/nova/conf/spice.py b/nova/conf/spice.py
index 8742da4..6cc45ae 100644
--- a/nova/conf/spice.py
+++ b/nova/conf/spice.py
@@ -225,6 +225,18 @@ Possible values:

 * False: concurrent access is not allowed.
 * True: concurrent access is allowed.
+"""),
+    cfg.BoolOpt('debug_logging',
+        default=False,
+        help="""
+Whether to emit SPICE debug logs or not to the qemu log.
+
+These debug logs are verbose, but can help diagnose some connectivity issues.
+
+Possible values:
+
+* False: no debug logs are enabled for SPICE connections.
+* True: debug logs are enabled for SPICE connections.
 """),
 ]

diff --git a/nova/tests/unit/virt/libvirt/test_driver.py b/nova/tests/unit/virt/libvirt/test_driver.py
index 4b02df8..eb11f28 100644
--- a/nova/tests/unit/virt/libvirt/test_driver.py
+++ b/nova/tests/unit/virt/libvirt/test_driver.py
@@ -6104,6 +6104,49 @@ class LibvirtConnTestCase(test.NoDBTestCase,
              '  <qemu:env name="SPICE_DEBUG_ALLOW_MC" value="1"/>\n'
              '</qemu:commandline>'))

+    def test_get_guest_config_with_spice_debug(self):
+        self.flags(enabled=False, group='vnc')
+        self.flags(virt_type='kvm', group='libvirt')
+        self.flags(enabled=True,
+                   agent_enabled=False,
+                   require_secure=True,
+                   allow_concurrent=False,
+                   debug_logging=True,
+                   server_listen='10.0.0.1',
+                   group='spice')
+        self.flags(pointer_model='usbtablet')
+
+        cfg = self._get_guest_config_with_graphics()
+        print(cfg.to_xml())
+        self.assertEqual(
+            cfg.qemu_commandline.to_xml().lstrip().rstrip(),
+            ('<qemu:commandline xmlns:qemu='
+                '"http://libvirt.org/schemas/domain/qemu/1.0">\n'
+             '  <qemu:env name="G_MESSAGES_DEBUG" value="all"/>\n'
+             '</qemu:commandline>'))
+
+    def test_get_guest_config_with_spice_concurrent_and_debug(self):
+        self.flags(enabled=False, group='vnc')
+        self.flags(virt_type='kvm', group='libvirt')
+        self.flags(enabled=True,
+                   agent_enabled=False,
+                   require_secure=True,
+                   allow_concurrent=True,
+                   debug_logging=True,
+                   server_listen='10.0.0.1',
+                   group='spice')
+        self.flags(pointer_model='usbtablet')
+
+        cfg = self._get_guest_config_with_graphics()
+        print(cfg.to_xml())
+        self.assertEqual(
+            cfg.qemu_commandline.to_xml().lstrip().rstrip(),
+            ('<qemu:commandline xmlns:qemu='
+                '"http://libvirt.org/schemas/domain/qemu/1.0">\n'
+             '  <qemu:env name="SPICE_DEBUG_ALLOW_MC" value="1"/>\n'
+             '  <qemu:env name="G_MESSAGES_DEBUG" value="all"/>\n'
+             '</qemu:commandline>'))
+
     @mock.patch.object(host.Host, 'get_guest')
     @mock.patch.object(libvirt_driver.LibvirtDriver,
                        '_get_serial_ports_from_guest')
diff --git a/nova/virt/libvirt/config.py b/nova/virt/libvirt/config.py
index de85c34..4ab3602 100644
--- a/nova/virt/libvirt/config.py
+++ b/nova/virt/libvirt/config.py
@@ -3291,6 +3291,9 @@ class LibvirtConfigGuest(LibvirtConfigObject):
     def set_commandline_spice_concurrent(self):
         self.qemu_commandline.spice_concurrent = True

+    def set_commandline_spice_debug(self):
+        self.qemu_commandline.spice_debug = True
+

 class LibvirtConfigGuestSnapshot(LibvirtConfigObject):

@@ -3930,6 +3933,7 @@ class LibvirtConfigGuestQemuCommandLine(LibvirtConfigObject):
                              ns_uri=QEMU_NS)

         self.spice_concurrent = False
+        self.spice_debug = False

     def format_dom(self):
         meta = super(LibvirtConfigGuestQemuCommandLine, self).format_dom()
@@ -3942,6 +3946,13 @@ class LibvirtConfigGuestQemuCommandLine(LibvirtConfigObject):
             meta.append(pkg)
             required = True

+        if self.spice_debug:
+            pkg = self._new_node('env')
+            pkg.set('name', 'G_MESSAGES_DEBUG')
+            pkg.set('value', 'all')
+            meta.append(pkg)
+            required = True
+
         if not required:
             return None
         return meta
diff --git a/nova/virt/libvirt/driver.py b/nova/virt/libvirt/driver.py
index da6ff6f..eeb5853 100644
--- a/nova/virt/libvirt/driver.py
+++ b/nova/virt/libvirt/driver.py
@@ -7410,6 +7410,8 @@ class LibvirtDriver(driver.ComputeDriver):
             graphics.secure = CONF.spice.require_secure
             if CONF.spice.allow_concurrent:
                 guest.set_commandline_spice_concurrent()
+            if CONF.spice.debug_logging:
+                guest.set_commandline_spice_debug()
             guest.add_device(graphics)
             add_video_driver = True

